### PR TITLE
Create es_query_executor.py

### PR DESCRIPTION
**Pull request by riverma**
## Purpose
- A script to satisfy the workflow described in https://github.com/nasa/opera-sds-ops/issues/19#issuecomment-1421278734 - basically to more systematically run specific elasticsearch queries in a controlled, log-conscious way for OPS support, including delete queries.
- Script does the following:
  - Takes in a JSON file that contains an Elasticsearch query and takes in a given Elasticsearch host and index name on the command-line
  - Executes the JSON within the input file as a query to Elasticsearch
  - Logs the output to a rolling log file
  - Prints the document IDs that were affected by the query to the command line
## Proposed Changes
- [ADD] New script 
## Issues
- https://github.com/nasa/opera-sds-ops/issues/19 
## Testing
- Static queries and script tested against a local Elasticsearch instance with sample docs identical to our OPERA Mozart / GRQ ESes. 
- No significant impact on ES instances when tested on development venue by measuring query response times and CPU impact on Mozart and GRQ machines measured (see post in comments)

### CODE REVIEW

**Comment by riverma** [comment]
@niarenaw - can you test this script out on one of our test clusters against the Elasticsearch policies described? See attached README for usage directions. I've encoded your sample deletion policies as configuration files specified in `queries` sub-folder.

**Comment by riverma** [comment]
@LalaP @niarenaw - let's discuss if the cleanup policies specified within the `queries` sub-folder need iteration. See README section "Features" for the policies currently implemented.

**Comment by riverma** [comment]
@philipjyoon - one pre-requisite for running / testing this script is the Python Elasticsearch Client. Do we have this available on any of our clusters' Mozart for @niarenaw to be able to test this script?

**Comment by hookhua** [comment]
The NISAR and SWOT SDS teams added new capabilities to HySDS Core release version 5 where some of the features in this ticket may no longer be needed. The newer release v5 has support for automatic rollover of old jobs from Mozart’s Elasticsearch via ES ILM rollover policy. See updated the documentation here for more info: https://hysds-core.atlassian.net/wiki/spaces/HYS/pages/2050686977/Customizing+Rollover+of+old+jobs+in+Mozart

**Comment by riverma** [comment]
> The NISAR and SWOT SDS teams added new capabilities to HySDS Core release version 5 where some of the features in this ticket may no longer be needed. The newer release v5 has support for automatic rollover of old jobs from Mozart’s Elasticsearch via ES ILM rollover policy. See updated the documentation here for more info: https://hysds-core.atlassian.net/wiki/spaces/HYS/pages/2050686977/Customizing+Rollover+of+old+jobs+in+Mozart

Thanks @hookhua - great point! We reviewed the HySDS upcoming features and concluded we'd use them for a future release of OPERA SDS, when we adopt HySDS V5. We'll have to use our own home-grown approach for supporting interim needs. That being said, we can use a roll-off policy for routinely cleaning out the Metrics Elasticsearch, since the indices involved are already using date-time tagged names.

**Comment by riverma** [comment]
Some notes from @niarenaw's testing of this script:
- We _do_ want to use deprecated `body=query` syntax for the current PCM that was reverted [here](https://github.com/nasa/opera-sds-ops/pull/22/commits/0eb4246604e218bcd60c3b111b13f4a73b80939d)
- The `search` feature is currently limited to returning 10k results since we do not support paging of results
- All the `queries/*.json` files are currently using time ranges with the `lt` (less than) parameter, whereas they should use `gt` (greater than)
- It would be helpful to somehow embed the index names for the queries inside the JSON files, so that we don't need to look that up

**Comment by niarenaw** [comment]
Functionality request:
- need to include HLS l30 and s30 param files 
A couple operability requests:
- parameterize the log file output dir
- make the info logged a little more readable

File: es_query_executor/queries/job_status-current/jobs_nominal_old.json [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+                }
+              },
+              {
+                "match": {
+                  "status": "job-deduped"
+                }
+              },
+              {
+                "match": {
+                  "status": "job-failed"
```
Comment (riverma) [inline es_query_executor/queries/job_status-current/jobs_nominal_old.json:25]: We don't want job-failed within this nominal deletion policy. We already have jobs_failed_old.json for the failed jobs.

File: es_query_executor/README.md [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+## Features
+
+* Contains static set of pre-configured Elasticsearch queries for the HySDS system
+* Executes static queries against a specified Elasticsearch instance, either in "count" mode (find docs) or "delete" mode (delete docs)
+* Logs the transaction to a log file (daily logs up to max-14 days rolling)
+* Prints the number of documents affected by the query
+  
+This script includes a number of pre-configured Elasticsearch queries described below. Note the folder structure: `queries/[index_name]/*.json`. **This folder structure is important: the `index_name` is extracted from the folder in the path to the query JSON file.** This design prevents mistakes of running a delete query against the wrong index.
+- [queries/job_status-current/jobs_nominal_old.json](queries/job_status-current/jobs_nominal_old.json) - (Mozart) if status in {completed, revoked, deduped, failed} AND creation timestamp > 14 days old
```
Comment (riverma) [inline es_query_executor/README.md:26]: Shouldn't include 'failed' job status, we already have that condition below.

**Comment by riverma** [comment]
## Testing on a live FWD processing SDS/PCM replica cluster:

### Summary

> No significant time taken for deletion policies on FWD test system containing 45 days worth of operational data, and no significant CPU impact on Mozart or GRQ machine to execute deletion operations via Elasticsearch. 

### Elasticsearch response times when executing delete queries

- queries/grq_v1.1.10_triaged_job/old.json
  - 0.016s
  - 86 documents deleted
- queries/grq_v2.0_l2_hls_s30/old.json
  - 0.203s
  - 7213 documents deleted
- queries/grq_v1.0_l3_dswx_hls/old.json
  - 0.457s
  - 10396 documents deleted
- queries/job_status-current/jobs_nominal_old.json
  - 2.154s
  - 41766 documents deleted
- queries/job_status-current/jobs_failed_old.json
  - 0.008s
  - 7 documents deleted
- queries/grq_v2.0_l2_hls_l30/old.json
  - 0.097s
  - 3184 documents deleted
- queries/task_status-current/resources_old.json
  - 3.042s
  - 0 documents deleted

### CPU load (maximum CPU spikes) 

NOTE: time to deletions is around 05-26T00:00:02 UTC

Screenshot 2023-05-26 at 4 22 03 PM

**Review by niarenaw — APPROVED** [review]
Looks good! I think some of the queries for GRQ can be simplified a bit since each of the indices only holds the relevant documents, but I doubt that has any real impact on performance.

**Review by niarenaw — APPROVED** [review]
Looks good! I think some of the queries for GRQ can be simplified a bit since each of the indices only holds the relevant documents, but I doubt that has any real impact on performance.

**Review by niarenaw — APPROVED** [review]
Looks good! I think some of the queries for GRQ can be simplified a bit since each of the indices only holds the relevant documents, but I doubt that has any real impact on performance.

**Comment by LalaP** [comment]
**so this is what we have at the moment.** 
[queries/job_status-current/jobs_nominal_old.json](https://github.com/nasa/opera-sds-ops/blob/issue-19/es_query_executor/queries/job_status-current/jobs_nominal_old.json) - (Mozart) if status in {completed, revoked, deduped} AND creation timestamp > **14 days old**

[queries/job_status-current/jobs_failed_old.json](https://github.com/nasa/opera-sds-ops/blob/issue-19/es_query_executor/queries/job_status-current/jobs_failed_old.json) - (Mozart) if status = failed AND creation timestamp > **30 days old**

[queries/task_status-current/resources_old.json](https://github.com/nasa/opera-sds-ops/blob/issue-19/es_query_executor/queries/task_status-current/resources_old.json) - (Mozart) if resource in {task, event, worker} AND creation timestamp > **7 days old**

[queries/grq_v1.0_l3_dswx_hls/old.json](https://github.com/nasa/opera-sds-ops/blob/issue-19/es_query_executor/queries/grq_v1.0_l3_dswx_hls/old.json) - (GRQ) if dataset in {L2_HLS_L30, L2_HLS_S30} AND creation timestamp > **14 days old**

[queries/grq_v2.0_l2_hls_l30/old.json](https://github.com/nasa/opera-sds-ops/blob/issue-19/es_query_executor/queries/grq_v2.0_l2_hls_l30/old.json) - (GRQ) if dataset L2_HLS_L30 AND creation timestamp > **14 days old**

[queries/grq_v2.0_l2_hls_s30/old.json](https://github.com/nasa/opera-sds-ops/blob/issue-19/es_query_executor/queries/grq_v2.0_l2_hls_s30/old.json) - (GRQ) if dataset L2_HLS_S30 AND creation timestamp > **14 days old**

[queries/grq_v1.1.10_triaged_job/old.json](https://github.com/nasa/opera-sds-ops/blob/issue-19/es_query_executor/queries/grq_v1.1.10_triaged_job/old.json) - (GRQ) triaged jobs dataset AND creation timestamp > **30 days old**

Since we are in earlier phases of mission I would suggest to keep GRQ on 30 days rolling window.  
After we are more stable. In other word when discovered bugs specially with interfaces is resolved then we can narrow down the rolling window considering that we will have another PGE and pipeline added to our system.
So if possible lets change the following as: 


[queries/grq_v1.0_l3_dswx_hls/old.json](https://github.com/nasa/opera-sds-ops/blob/issue-19/es_query_executor/queries/grq_v1.0_l3_dswx_hls/old.json) - (GRQ) if dataset in {L2_HLS_L30, L2_HLS_S30} AND creation timestamp > **30 days old**

[queries/grq_v2.0_l2_hls_l30/old.json](https://github.com/nasa/opera-sds-ops/blob/issue-19/es_query_executor/queries/grq_v2.0_l2_hls_l30/old.json) - (GRQ) if dataset L2_HLS_L30 AND creation timestamp > **30 days old**

[queries/grq_v2.0_l2_hls_s30/old.json](https://github.com/nasa/opera-sds-ops/blob/issue-19/es_query_executor/queries/grq_v2.0_l2_hls_s30/old.json) - (GRQ) if dataset L2_HLS_S30 AND creation timestamp > **30 days old**

**Pull request merged**